### PR TITLE
updated release process

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: pages/release-process.md [hunk 1]
```diff
@@ -1,111 +1,113 @@
...
+layout: page
+title: Release Process
+permalink: /release-process/
+---
+
+### Initial Setup
+
+Before you can release Fluo or Fluo Recipes, you will need a GPG key. For information on generating
+a key look at this [ASF GPG page](https://www.apache.org/dev/openpgp.html).  After generating a key,
+add it to the [KEYS](https://www.apache.org/dist/incubator/fluo/KEYS) file.
```
Comment (ctubbsii) [inline pages/release-process.md]: Could mention that instructions are provided at top of KEYS file on how to do this.

File: pages/release-process.md [hunk 2]
```diff
@@ -1,111 +1,113 @@
...
+The maven release plugin will need credentials to stage artifacts.  You can provide the credentials
+by adding the following to your `~/.m2/settings.xml` file.  You may want remove your password after
+running maven.
+
+```xml
+<servers>
+  <server>
+    <id>apache.releases.https</id>
+    <username>your-apache-id</username>
+    <password>your-apache-password</password>
```
Comment (ctubbsii) [inline pages/release-process.md]: Should also link to Maven page on securing your credentials: https://maven.apache.org/guides/mini/guide-encryption.html

File: pages/release-process.md [hunk 3]
```diff
@@ -1,111 +1,113 @@
...
+  </server>
+</servers>
+```
+
+### Release Fluo
+
+Before starting the release process below, the following tasks should be complete:
+
+ * Create release notes for project website using GitHub issues.
```
Comment (ctubbsii) [inline pages/release-process.md]: Make bullet points consistent with terminating punctuation.
* Some lines have a period.
* Some don't

File: pages/release-process.md [hunk 4]
```diff
@@ -1,111 +1,113 @@
...
+ * Perform testing and document results
+ * Start a gpg-agent to cache your gpg key to avoid entering your passphrase multiple times.
+
+   ```shell
+   gpg-connect-agent reloadagent /bye
```
Comment (ctubbsii) [inline pages/release-process.md]: I don't think this is the correct command to start the gpg-agent. The previously documented command actually starts a gpg-agent. This may not be necessary if you're running inside a desktop environment, like GNOME or KDE, which provides a gpg-agent for you (gnome-keyring-daemon provides this in GNOME, for example).

What this gpg-connect-agent command does is clears the currently running gpg-agent's cache. The create-release-candidate script does this to help guarantee that the timer for your cached passphrase is reset before a potentially long-running build. The user still must ensure that their gpg-agent is configured with a sufficiently long timeout so that the cached passphrase won't expire during the build.

File: pages/release-process.md [hunk 5]
```diff
@@ -1,111 +1,113 @@
...
+   ```
+
+Next, repeat the steps below until a good release candidate (RC) is found.  The script in
+`contrib/create-release-candidate.sh` automates this process.  However there is no guarantee that it
+works correctly.  Before using the script ensure you understand the process and inspect the script.
+In the following steps `RCV` is short for release candidate version.
```
Comment (ctubbsii) [inline pages/release-process.md]: Might be worth mentioning that `0` is a good choice for the `RCV` for testing the release candidate process itself or for branching a pre-release candidate for testing. It's not a strictly necessary convention, but I've found it to be a useful one to make sure I don't get confused between an actual release candidate and a test one, not intended for voting.

File: pages/release-process.md [hunk 6]
```diff
@@ -1,111 +1,113 @@
...
+
+ 1. Branch master (or the current snapshot) and call the branch `<releaseVersion>-rc<RCV>-next`
+
+ 2. Except in Maven POMs (which will be updated by the next step), change any version references in docs and code from `<releaseVersion>-SNAPSHOT` to `<releaseVersion>`.  Commit changes to branch.
```
Comment (ctubbsii) [inline pages/release-process.md]: This really shouldn't be necessary. If we ever have to do this, it means we haven't done a good enough job with the build to auto-substitute these version numbers based on the current build. If there exist any such cases currently, we should file issues to automate them. This step should never be necessary during release candidate creation.

File: pages/release-process.md [hunk 7]
```diff
@@ -1,111 +1,113 @@
...
+
+ 3. Prepare the release which will verify that all tests pass: `mvn release:prepare`
```
Comment (ctubbsii) [inline pages/release-process.md]: This step also creates two commits in the current branch, and creates a tag of the first commit of these two. That tag is used by `release:prepare`, but should be deleted afterwards.

File: pages/release-process.md [hunk 8]
```diff
@@ -1,111 +1,113 @@
...
+
+ 4. Perform the release: `mvn release:perform`
+    * This step will create a staging repository viewable only to you when login to https://repository.apache.org
```
Comment (ctubbsii) [inline pages/release-process.md]: The staging repository is viewable by all committers on the project, not just its creator. Also, it's useful to put in a description of the repo when one closes it. Something like "Apache Fluo (incubating) 1.1.0-rc3".

File: pages/release-process.md [hunk 9]
```diff
@@ -1,111 +1,113 @@
...
+    * When `release:perform` finishes, you will need to close the staging repository to make it viewable to anyone at  https://repository.apache.org/content/repositories/orgapachefluo-REPO_ID
```
Comment (ctubbsii) [inline pages/release-process.md]: It's important that the staging repo is *never* promoted/released until after a vote is successful, because that will push its contents to Maven central as though it were an official release. This can't be undone, and the only correct course of action (if the vote fails) is to skip over that release number for the next release candidate.

File: pages/release-process.md [hunk 10]
```diff
@@ -1,111 +1,113 @@
...
+ 5. Delete the tag created by Maven in previous step.  This tag should not be pushed to Apache until the vote passes.  Also, a signed tag should be created instead of the one created by Maven.  So out of an abundance of caution its best to delete it now and create the signed tag after the vote passes.
+
+    ```shell
+    git tag -D rel/fluo-<releaseVersion>
+    ```
+
+ 6. Push the `<releaseVersion>-rc<RCV>-next` to apache.
+
+    ```shell
+    git checkout <releaseVersion>-rc<RCV>-next
```
Comment (ctubbsii) [inline pages/release-process.md]: This should already be the current checkout at this point.

File: pages/release-process.md [hunk 11]
```diff
@@ -1,111 +1,113 @@
...
+    git push apache-remote <releaseVersion>-rc<RCV>-next
+    ```
+
+ 7. Create the release candidate branch `<releaseVersion>-rc<RCV>` and push it.  This branch should be one commit behind `<releaseVersion>-rc<RCV>-next`
```
Comment (ctubbsii) [inline pages/release-process.md]: ... and one commit ahead of the branch point.

File: pages/release-process.md [hunk 12]
```diff
@@ -1,111 +1,113 @@
...
+    ```shell
+    git checkout -b <releaseVersion>-rc<RCV> <releaseVersion>-rc<RCV>-next~1
+    git push -u apache-remote <releaseVersion>-rc<RCV>
+    ```
+
+ 8. Send a message to the devs to let them know a release is staged. This [example][example-email] for the Fluo 1.0.0 release can be used as template.  The script `contrib/create-release-candidate.sh` can be used to generate this email.
+
+When the vote passes on a release candidate, follow the steps below to complete the release using the chosen RC:
+
+ 1. Merge your RC branch into master and push those commits upstream.  Afterwards, you can delete your RC branch.
```
Comment (ctubbsii) [inline pages/release-process.md]: The "-next" branch should be merged into (or become, if one didn't already exist) the maintenance branch for the version which was just released (if a maintenance branch is desired), and then that maintenance branch should be merged forward into maintenance branches for newer releases, until eventually merging into master. (not sure how to make that more succinct)

File: pages/release-process.md [hunk 13]
```diff
@@ -1,111 +1,113 @@
...
+
+    ```shell
+    git checkout master
+    git merge <releaseVersion>-rc<RCV>-next
+    git push apache-remote master
+    ```
+
+ 2. Release the artifacts at https://repository.apache.org so that they get published in Maven Central.  You can drop any staging repos for RCs that were not chosen.
```
Comment (ctubbsii) [inline pages/release-process.md]: Again, useful to have a comment: "Vote failed for Apache Fluo 1.1.0-rc3" or similar for dropping old staging repos; "Apache Fluo 1.1.0" for released repos.

File: pages/release-process.md [hunk 14]
```diff
@@ -1,111 +1,113 @@
...
+
+ 3. Create a signed tag for the release from the chosen RC tag and push to upstream repo:
+
+    ```shell
+    # Create signed tag from RC2 tag.
+    # You may need to use -u <key-id> to specify GPG key
+    git tag  -f -m 'Apache Fluo <releaseVersion>' -s rel/fluo-<releaseVersion> <releaseVersion>-rc<RCV>
```
Comment (ctubbsii) [inline pages/release-process.md]: Users may wish to do `git config --global --get gpg.program || git config --global --add gpg.program gpg2`
This is because some systems which have both `gpg` and `gpg2` installed may not function correctly. (See https://bugzilla.redhat.com/show_bug.cgi?id=568406)

File: pages/release-process.md [hunk 15]
```diff
@@ -1,111 +1,113 @@
...
+    # Verify the tag is the expected commit
+    git log -1 rel/fluo-<releaseVersion>
+    # Push signed tag to upstream repo
+    git push apache-remote rel/fluo-<releaseVersion>
```
Comment (ctubbsii) [inline pages/release-process.md:127]: Are you sure this is correct? I thought you needed the `--tags` option to push tags unless you specify the fully-qualified ref name: `git push origin refs/tags/release-1.0:refs/tags/release-1.0`

I think what your command will do is try to push a new branch from the currently checked-out branch/commit to the remote with the name `rel/fluo-<releaseVersion>`.

Might be a good idea to document testing it with `--dry-run` first.

File: pages/release-process.md [hunk 16]
```diff
@@ -1,111 +1,113 @@
...
+    git push apache-remote --delete <releaseVersion>-rc<RCV>-next
+    git branch -d <releaseVersion>-rc<RCV>-next
+    git push apache-remote --delete <releaseVersion>-rc<RCV>
+    git branch -d <releaseVersion>-rc<RCV>
+    ```
+ 6.  View the [website README] for instructions on how to generate Javadocs and documentation using the released tag.  Submit PR to the website repo to publish.
+
+ 7.  Send an email to `dev@fluo.incubator.apache.org` announcing new release.
+
+[website README]: https://github.com/apache/incubator-fluo-website/blob/master/README.md
```
Comment (ctubbsii) [inline pages/release-process.md]: I didn't know spaces worked in markdown link references!

**Review by ctubbsii — CHANGES_REQUESTED** [review]
This is good stuff. This page was *way* out of date. I made a few suggestions.

File: pages/release-process.md (hunk 10 above)
Comment (keith-turner) [reply pages/release-process.md]: I was just making it crystal clear.

File: pages/release-process.md (hunk 15 above)
Comment (keith-turner) [reply pages/release-process.md:127]: I think I tried this pushing to my fork, but not completely sure.  I'll test again.

File: pages/release-process.md (hunk 8 above)
Comment (ctubbsii) [reply pages/release-process.md]: In the latest version, this wording is still incorrect.

File: pages/release-process.md [hunk 17]
```diff
@@ -1,111 +1,145 @@
...
+
+ 5. Push the `<releaseVersion>-rc<RCV>-next` to apache.
+
+    ```shell
+    git checkout <releaseVersion>-rc<RCV>-next
+    git push apache-remote <releaseVersion>-rc<RCV>-next
+    ```
+
+ 6. Create the release candidate branch `<releaseVersion>-rc<RCV>` and push it.  This branch should
+    be one commit behind `<releaseVersion>-rc<RCV>-next` and one commit behind the branch point.
```
Comment (ctubbsii) [inline pages/release-process.md]: ahead of the branch point :)

File: pages/release-process.md [hunk 18]
```diff
@@ -1,111 +1,145 @@
...
+In the following steps `RCV` is short for release candidate version.  For the case where you want to
+make an initial release candidate available for evaluation, but nor for voting, consider using `0`
+for `RCV`.
+
+ 1. Branch master (or the current snapshot) and call the branch `<releaseVersion>-rc<RCV>-next`
+
+ 2. Prepare the release which will verify that all tests pass: `mvn release:prepare`
+
+ 3. Perform the release: `mvn release:perform`
+    * This step will create a staging repository viewable only to you when login to https://repository.apache.org
```
Comment (ctubbsii) [inline pages/release-process.md]: More accurate:
s/only to you/to you only/

It's visible to all committers on the project after logging in, but it's not necessary to specify all that. Shifting the word "only" to the right twice makes the statement correct.

**Comment by ctubbsii** [comment]
Recent changes look good. I see you incorporated pretty much all of my feedback from my first review. Thanks for that! I reiterated two more minor issues from my first review that I think should be addressed, but otherwise, looks good to me.

File: pages/release-process.md (hunk 18 above)
Comment (keith-turner) [reply pages/release-process.md]: I decided to drop the whole bit.  Even it we make it correct, I feel its not very helpful/useful info.

**Comment by keith-turner** [comment]
Still need to document putting the source and bin tarballs in svn so they end up on the mirrors.  Also need to document what needs to be done to remove older releases from mirrors.

File: pages/release-process.md (hunk 15 above)
Comment (keith-turner) [reply pages/release-process.md:127]: I used this command to create https://github.com/apache/incubator-fluo/releases/tag/rel%2Ffluo-1.1.0-incubating

**Review by ctubbsii — APPROVED** [review]
Everything looks good to me. Anything new should probably start in a new PR. This one is getting hard to follow with all the incremental updates, especially when comparing to the original, due to the dos2unix line ending change.
